session.go

```diff
@@ -103,33 +110,50 @@ func DefaultSessionHandler(srv *Server, conn *gossh.ServerConn, newChan gossh.Ne
 		sessReqCb:         srv.SessionRequestCallback,
 		subsystemHandlers: srv.SubsystemHandlers,
 		ctx:               ctx,
+		wroteCh:           make(chan interface{}),
 	}
 	sess.handleRequests(reqs)
 }
 
 type session struct {
 	sync.Mutex
 	gossh.Channel
-	conn              *gossh.ServerConn
-	handler           Handler
-	subsystemHandlers map[string]SubsystemHandler
-	handled           bool
-	exited            bool
-	pty               *Pty
-	winch             chan Window
-	env               []string
-	ptyCb             PtyCallback
-	sessReqCb         SessionRequestCallback
-	rawCmd            string
-	subsystem         string
-	ctx               Context
-	sigCh             chan<- Signal
-	sigBuf            []Signal
-	breakCh           chan<- bool
+	conn                *gossh.ServerConn
+	handler             Handler
+	subsystemHandlers   map[string]SubsystemHandler
+	handled             bool
+	exited              bool
+	pty                 *Pty
+	winch               chan Window
+	env                 []string
+	ptyCb               PtyCallback
+	sessReqCb           SessionRequestCallback
+	rawCmd              string
+	subsystem           string
+	ctx                 Context
+	sigCh               chan<- Signal
+	sigBuf              []Signal
+	breakCh             chan<- bool
+	disablePtyEmulation bool
+
+	// wroteCh is closed after the first call to Write.
+	wroteCh          chan interface{}
+	wroteChCloseOnce sync.Once
+}
+
+func (sess *session) DisablePTYEmulation() error {
+	select {
+	case <-sess.wroteCh:
+		return errors.New("DisablePTYEmulation must be called before calls to Write")
+	default:
+	}
+	sess.disablePtyEmulation = true
+	return nil
 }
 
 func (sess *session) Write(p []byte) (n int, err error) {
-	if sess.pty != nil {
+	sess.wroteChCloseOnce.Do(func() { close(sess.wroteCh) })
```

Comment:
pretty sure this allocs once per Write. You'd need to use a method value to avoid that last I checked.

Reply:
Removed this with the error